crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -652,6 +519,293 @@ where
+            stats,
+            changed_keys,
+        }: TxInfo,
+    ) {
+        match tx_result {
+            Ok(tx_result) => {
+                // Track the need to commit the batch hash for replay
+                // protection. Hash must be written if at least one of the
+                // txs in the batch requires so
+                let mut commit_batch_hash = false;
+                // Atomic batches need custom handling when even a single tx
+                // fails, since we need to drop everything
+                if is_atomic_batch
+                    && tx_result.batch_results.0.iter().any(|(_, result)| {
+                        match result {
+                            Ok(res) => {
+                                // If an inner tx failed for any reason but
+                                // invalid signature, commit its hash to
+                                // storage,
+                                // otherwis allow for a replay
+                                if !res
+                                    .vps_result
+                                    .status_flags
+                                    .contains(VpStatusFlags::INVALID_SIGNATURE)
+                                {
+                                    commit_batch_hash = true;
+                                }
+
+                                !res.is_accepted()
+                            }
+                            Err(e) => {
+                                // If inner transaction didn't fail
+                                // because of invalid
+                                // section commitment, commit its hash to
+                                // prevent replays
+                                if matches!(
+                                    tx_header.tx_type,
+                                    TxType::Wrapper(_)
+                                ) && !matches!(
+                                    e,
+                                    protocol::Error::MissingSection(_)
+                                ) {
+                                    commit_batch_hash = true;
+                                }
+                                true
+                            }
+                        }
+                    })
+                {
+                    // TODO: nee to increment the rejected txs too
+                    for _ in 0..commitments_len {
+                        stats.increment_errored_txs();
+                    }
+                    self.state.write_log_mut().drop_batch();
+                    tx_event.extend(Code(ResultCode::InvalidTx));
+                } else {
+                    for (cmt_hash, batched_result) in &tx_result.batch_results.0
+                    {
+                        match batched_result {
+                            Ok(result) => {
+                                if result.is_accepted() {
+                                    if is_masp_tx(&result.changed_keys) {
```

Comment:
if possible, we should move this branch to another helper method. it gets kinda confusing to read, given the many indentation levels

Reply:
I've reworked the entire `evaluate_tx_result` function